### PR TITLE
build: Use strict eslint config in common/definitions packages

### PR DESCRIPTION
**Pull request by tylerbutler**
Updates several common packages to use the strict eslint config, primarily so type-only exports/imports are enforced.

### CODE REVIEW

File: packages/common/container-definitions/src/fluidPackage.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@
  * Specifies an environment on Fluid property of a IFluidPackage.
  * @alpha
  */
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
```
Comment (tylerbutler) [inline packages/common/container-definitions/src/fluidPackage.ts:10]: disabled so the API doesn't change - not sure there's any difference though.

Comment (jason-ha) [reply packages/common/container-definitions/src/fluidPackage.ts:10]: There seem to be enough of these that it seems like step 1 is to use strict but roll this one back.
A lot of single line disables is a could be preferred or not. In the case of not really knowing, I like not single lines.

Another good rule is that all disables have a comment, saying why.
I didn't look this one rule up in particular. Just commenting on the general pattern.

File: packages/common/container-definitions/src/audience.ts [hunk 2]
```diff
@@ -5,8 +5,8 @@
 
 // False positive: this is an import from the `events` package, not from Node.
 // eslint-disable-next-line unicorn/prefer-node-protocol
-import { EventEmitter } from "events";
-import { IClient } from "@fluidframework/protocol-definitions";
+import { type EventEmitter } from "events";
+import { type IClient } from "@fluidframework/protocol-definitions";
```
Comment (jason-ha) [inline packages/common/container-definitions/src/audience.ts:10]: Is there a tool that is choosing to resolve these with per type `type` imports versus `type` import the whole block (`import type {`)?
I noticed that index.ts imports are using whole block which is the most important IMO. But I like whole block as default so I can see that there really isn't a runtime dep.
If there is a tool with rationale, then I do prefer automation. :)

File: packages/common/driver-definitions/src/urlResolver.ts [hunk 3]
```diff
@@ -92,6 +94,6 @@ export interface IDriverHeader {
 }
 
 declare module "@fluidframework/core-interfaces" {
-	// eslint-disable-next-line @typescript-eslint/no-empty-interface
+	// eslint-disable-next-line @typescript-eslint/no-empty-interface, jsdoc/require-jsdoc
```
Comment (jason-ha) [inline packages/common/driver-definitions/src/urlResolver.ts:97]: >, jsdoc/require-jsdoc [](http://example.com/codeflow?start=66&length=21)

Shouldn't be doced?

File: packages/common/container-definitions/src/audience.ts (hunk 2 above)
Comment (tylerbutler) [reply packages/common/container-definitions/src/audience.ts:10]: I used the eslint rule's auto-fixer, but I also did some manual tweaks. These are the relevant config settings we currently have:

```json
        "@typescript-eslint/consistent-type-exports": [
            "error"
        ],
        "@typescript-eslint/consistent-type-imports": [
            "error",
            {
                "fixStyle": "inline-type-imports"
            }
        ],
```

Looks like there's another setting we could use: https://typescript-eslint.io/rules/consistent-type-imports/#fixstyle

> I noticed that index.ts imports are using whole block which is the most important IMO. But I like whole block as default so I can see that there really isn't a runtime dep.

You still have to be careful to check for a second non-type-only export from the same module, right?

Comment (jason-ha) [reply packages/common/container-definitions/src/audience.ts:10]: Yes, you would unless we ban double import, which I think is pretty common.
What manual tweaks were you making (broadly speaking)?

File: packages/common/driver-definitions/src/urlResolver.ts (hunk 3 above)
Comment (tylerbutler) [reply packages/common/driver-definitions/src/urlResolver.ts:95]: @kian-thompson @vladsud Do you know how we should document this?

File: packages/common/container-definitions/src/audience.ts (hunk 2 above)
Comment (tylerbutler) [reply packages/common/container-definitions/src/audience.ts:10]: Opened #19479 to update the lint config to prefer block-level type-only exports/imports

File: packages/common/driver-definitions/src/urlResolver.ts (hunk 3 above)
Comment (kian-thompson) [reply packages/common/driver-definitions/src/urlResolver.ts:95]: I'm not sure what this is. @jatgarg Do you have a clue?

Comment (alexvy86) [reply packages/common/driver-definitions/src/urlResolver.ts:95]: Also curious 👀

File: packages/common/core-interfaces/src/test/types/fluidObjectTypes.ts [hunk 4]
```diff
@@ -7,7 +7,8 @@
 /* eslint-disable @typescript-eslint/no-unsafe-call */
 /* eslint-disable @typescript-eslint/no-unsafe-member-access */
 
-import { IFluidLoadable, IProvideFluidLoadable, FluidObject, FluidObjectKeys } from "../../";
+import type { IProvideFluidLoadable, FluidObject, FluidObjectKeys } from "../../";
+import { IFluidLoadable } from "../../";
```
Comment (jason-ha) [inline packages/common/core-interfaces/src/test/types/fluidObjectTypes.ts:12]: Shouldn't this be a type import?

Comment (jason-ha) [reply packages/common/core-interfaces/src/test/types/fluidObjectTypes.ts:12]: nm - see the uses on lines 67,68

Comment (tylerbutler) [reply packages/common/core-interfaces/src/test/types/fluidObjectTypes.ts:12]: IFluidLoadable is a const.

```ts
export const IFluidLoadable: keyof IProvideFluidLoadable = "IFluidLoadable";
```

File: packages/common/driver-definitions/src/urlResolver.ts (hunk 3 above)
Comment (jason-ha) [reply packages/common/driver-definitions/src/urlResolver.ts:95]: Without traction can we file a work item to follow up (and add comment)? Then, I'd be happy to approve. :)

File: packages/common/core-interfaces/src/test/types/fluidObjectTypes.ts (hunk 4 above)
Comment (tylerbutler) [reply packages/common/core-interfaces/src/test/types/fluidObjectTypes.ts:12]: Right, I was just trying to explain why it couldn't be imported as a type only - but I guess the real answer is that it's USED as a const in this file, not that it is defined as both a const and a type.

File: packages/common/driver-definitions/src/urlResolver.ts (hunk 3 above)
Comment (tylerbutler) [reply packages/common/driver-definitions/src/urlResolver.ts:95]: I'd rather not continue to block this PR on this, so if there's no info by later today, I will add a tracking issue, note it in a comment, and get this merged.

**Review by jason-ha — APPROVED** [review]
(yay - know lets see how long the work item lingers)

**Pull request merged**